rfd/0003-deduplication.md

```diff
@@ -0,0 +1,214 @@
+
+#### Usage of started_at, endated_at, boefje and runnable_hash fields
+
+This isn't used by any boefje and can be safely removed.
+
+#### Boefje signature / arguments
+
+Boefje that does not require organization can have the following signature:
+
+```python
+def run(input_ooi: dict[str, Any]) -> list[tuple[set, bytes | str]]:
+```
+
+And boefje that requires organization:
+
+```python
+def run(input_ooi: dict[str, Any], organization: str) -> list[tuple[set, bytes | str]]:
+```
+
+The `input_ooi` is the serialized input OOI that is stored in the BoefjeMeta
+`arguments`. The environment field does need to be passed as argument because it
+is already passed to the boefje using environment variables.
+
+### Deduplication implementation
+
+The scheduler is the place where boefjes are scheduled so that seems to be most
+obvious place to implement deduplication. Because the data necessary is
+currently spread out over multiple services / databases this means it would need
+to do requests to other services to fetch this information. This is unavoidable
+because there is currently no single place with all the information.
+
+#### Whether boefje needs organization
+
+The scheduler already fetches and caches all the available boefjes for an
+organization, so the scheduler can use this data to easily check if the boefje
+requires the organization and can't be deduplicated.
+
+#### Environment settings from previous task in other organization
+
+This data is stored in bytes and needs to be fetched by the scheduler to
+determine whether the environment settings are the same.
+
+#### Environment settings of to be deduplicated boefje task
+
+This is a complicated. The scheduler can fetch the current configured
+environment from the katalogus, but this is merged with the environment that is
+passed directly to the boefje runner as environment variables. There is
+currently no easy way to fetch those and adding an API for this to the boefje
+runner doesn't seem to be a good idea.
+
+The best course of action seems to be to disallow the environment variables
+passed to the boefje runner when deduplication is turned on. If both
+deduplication is turned on and environment variables are passed to the boefje
+runner we need to return an error because the user needs to know that the passed
+variables aren't used.
+
+#### Deduplication algorithm
+
+A possible algorithm for deduplication could be:
+
+- The scheduler wants to scheduler a boefje for a specific organization and
+  input OOI.
+- The scheduler checks if the boefje needs the organization. If so it can
+  continu with scheduling the boefje task.
+- The scheduler queries the stored tasks to see if a different organization has
+  already run on this input OOI.
+- If there is a match, the scheduler fetches the environment of the previous
+  boefje task and the environment of the new boefje task. If they are the same
+  they can be deduplicated. If they differ the new boefje needs to be scheduled.
```

Comment:
This could be a nice solution, and from the scheduler point of view this can be implemented. A few questions:

1. When the scheduler deems the `Task` to be a duplication, do we still persist the `Task` for that organisation but its results to be linked to the results of another organisation? If so do we need to communicate this to the user, that their `Task` has been resolved by the results of another `Task` from another organisation?
2. The `BoefjeScheduler` as-is, is complex in relating to checking whether or not a `Task` should be pushed on the queue, do you have any ideas or suggestions for adding these suggestions while making sure that we don't add too much additional complexity?

Reply:
1. For an initial implementation I'd say the user doesn't need to notice anything, except maybe that the job was resolved faster than usual.